src/EvaluationFunctor.hpp

```diff
@@ -1,27 +1,116 @@
     Taylor and Interpolation tables can have the same interface
+  - easy (though obtuse) to use if mathematical functions are templated
+  - copy and paste the following example code into a new file and use
+    the command :%s/new_name/foo/g in vim or (TODO something else) in emacs 
+    to write your own function.
+
+  Example usage:
+  #include EvaluationFunctor.hpp
+  // this could just be a function when we move to std::function
+  template <typename T>
+  struct foo : EvaluationFunctor
+  {
+    T operator()(T x){ return x; }
+  };
+
+  int main(){
+    // these 'new' memory allocations can be removed when we move to std::function
+    FunctionContainer foo_container{new foo<double>, new foo<fvar1>,
+        new foo<fvar2>, new foo<fvar3>, new foo<fvar4>, 
+        new foo<fvar5>, new foo<fvar6>, new foo<fvar7>};
+    // or just
+    // FunctionContainer foo_container;
+    // foo_container.double_func = new foo<double>;
+    // if you're not using any Taylor/Pade/Hermite tables
+    // (the compiler will warn you)
+    return 0;
+  }
 */
 #pragma once
+#include <stdexcept>
+#include <memory>
+#include <boost/math/differentiation/autodiff.hpp>
+
+using boost::math::differentiation::autodiff_fvar;
+typedef autodiff_fvar<double,1> fvar1;
+typedef autodiff_fvar<double,2> fvar2;
+typedef autodiff_fvar<double,3> fvar3;
+typedef autodiff_fvar<double,4> fvar4;
+typedef autodiff_fvar<double,5> fvar5;
+typedef autodiff_fvar<double,6> fvar6;
+typedef autodiff_fvar<double,7> fvar7;
 
-// TODO use a better exception type
-#define NOT_DEFINED(ORDER)			\
-  throw ORDER" derivative not defined.";	\
-  return 0;
+/* Used by each table type to check if the required function
+ * type has been provided.
+ * TODO decide whether or not to make this null for -DNDEBUG */
+#define __IS_NULLPTR(VAR) \
+    if(VAR == nullptr)     \
+      throw std::invalid_argument(#VAR " is a nullptr")
 
 template <class IN_TYPE, class OUT_TYPE>
 class EvaluationFunctor
 {
 public:
   virtual OUT_TYPE operator()(IN_TYPE x) = 0;
-  /* Raise an exception in these by default? */
-  virtual OUT_TYPE deriv(IN_TYPE x){ NOT_DEFINED("1st"); };
-  virtual OUT_TYPE deriv2(IN_TYPE x){ NOT_DEFINED("2nd"); };
-  virtual OUT_TYPE deriv3(IN_TYPE x){ NOT_DEFINED("3rd"); };
-  virtual OUT_TYPE deriv4(IN_TYPE x){ NOT_DEFINED("4th"); };
-  virtual OUT_TYPE deriv5(IN_TYPE x){ NOT_DEFINED("5th"); };
-  virtual OUT_TYPE deriv6(IN_TYPE x){ NOT_DEFINED("6th"); };
-  virtual OUT_TYPE deriv7(IN_TYPE x){ NOT_DEFINED("7th"); };
+};
+
+
+class FunctionContainer
+{
+  // create a set of structs so we can specify function signatures
+  // (ie, number of differentiaions needed) with an index
+  template<unsigned int N>
+  struct func_type{
+    using type = EvaluationFunctor<autodiff_fvar<double,N>,autodiff_fvar<double,N>>*;
+  };
+  // special case for N=0
+  template<>
+  struct func_type<0>{
+    using type = EvaluationFunctor<double,double>*;
+  };
+
```

Comment:
Fails in g++ (Ubuntu 9.4.0-1ubuntu1~20.04.1) 9.4.0 with error:
`error: explicit specialization in non-namespace scope`
Apparently this is not standard-compliant, but is accepted by clang